File: samples/Windows/java/NonTableManagerSamples/TwoConnections/src/ResponseListener.java
package twoconnections;

import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

import com.fxcore2.*;

public class ResponseListener implements IO2GResponseListener {
    private O2GSession mSession;
    private String mRequestID;
    private O2GResponse mResponse;
    private String mOrderID;
    private Semaphore mSemaphore;

    // ctor
    public ResponseListener(O2GSession session) {
        mSession = session;
        mRequestID = "";
        mResponse = null;
        mOrderID = "";
        mSemaphore = new Semaphore(0);
    }

    public void setRequestID(String sRequestID) {
        mResponse = null;
        mRequestID = sRequestID;
    }

    public boolean waitEvents() throws Exception {
        return mSemaphore.tryAcquire(30, TimeUnit.SECONDS);
    }

    public O2GResponse getResponse() {
        return mResponse;
    }

    public String getOrderID() {
        return mOrderID;
    }

    public void onRequestCompleted(String sRequestId, O2GResponse response) {
        if (mRequestID.equals(response.getRequestId())) {
            mResponse = response;
            mSemaphore.release();
        }
    }

    public void onRequestFailed(String sRequestID, String sError) {
        if (mRequestID.equals(sRequestID)) {
            System.out.println("Request failed: " + sError);
            mSemaphore.release();
        }
    }

    public void onTablesUpdates(O2GResponse response) {
        O2GResponseReaderFactory factory = mSession.getResponseReaderFactory();
        if (factory != null) {
            O2GTablesUpdatesReader reader = factory.createTablesUpdatesReader(response);
            for (int ii = 0; ii < reader.size(); ii++) {
                if (reader.getUpdateTable(ii) == O2GTableType.ORDERS) {
                    O2GOrderRow orderRow = reader.getOrderRow(ii);
                    if (mRequestID.equals(orderRow.getRequestID())) {
                        if (reader.getUpdateType(ii) == O2GTableUpdateType.INSERT) {
                            mOrderID = orderRow.getOrderID();
                            System.out.println(String.format("The order has been added. OrderID=%s, Type=%s, BuySell=%s, Rate=%s, TimeInForce=%s",
                                    orderRow.getOrderID(), orderRow.getType(), orderRow.getBuySell(), orderRow.getRate(), orderRow.getTimeInForce()));
                            mSemaphore.release();
                            break;
                        }
                    }
                }
            }
        }
    }
}

File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/NonTableManagerSamples/RemoveOrder/src/ResponseListener.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using fxcore2;

namespace RemoveOrder
{
    class ResponseListener : IO2GResponseListener
    {
        private O2GSession mSession;
        private string mRequestID;
        private O2GResponse mResponse;
        private string mOrderID;
        private EventWaitHandle mSyncResponseEvent;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="session"></param>
        public ResponseListener(O2GSession session)
        {
            mRequestID = string.Empty;
            mOrderID = string.Empty;
            mResponse = null;
            mSyncResponseEvent = new EventWaitHandle(false, EventResetMode.AutoReset);
            mSession = session;
        }

        public void SetRequestID(string sRequestID)
        {
            mOrderID = string.Empty;
            mResponse = null;
            mRequestID = sRequestID;
        }

        public bool WaitEvents()
        {
            return mSyncResponseEvent.WaitOne(30000);
        }

        public O2GResponse GetResponse()
        {
            return mResponse;
        }

        public string GetOrderID()
        {
            return mOrderID;
        }

        #region IO2GResponseListener Members

        public void onRequestCompleted(string sRequestId, O2GResponse response)
        {
            if (mRequestID.Equals(response.RequestID))
            {
                mResponse = response;
            }
        }

        public void onRequestFailed(string sRequestID, string sError)
        {
            if (mRequestID.Equals(sRequestID))
            {
                Console.WriteLine("Request failed: " + sError);
                mSyncResponseEvent.Set();
            }
        }

        public void onTablesUpdates(O2GResponse data)
        {
            O2GResponseReaderFactory factory = mSession.getResponseReaderFactory();
            if (factory != null)
            {
                O2GTablesUpdatesReader reader = factory.createTablesUpdatesReader(data);
                for (int ii = 0; ii < reader.Count; ii++)
                {
                    if (reader.getUpdateTable(ii) == O2GTableType.Orders)
                    {
                        O2GOrderRow orderRow = reader.getOrderRow(ii);
                        if (orderRow.RequestID.Equals(mRequestID))
                        {
                            switch (reader.getUpdateType(ii))
                            {
                                case O2GTableUpdateType.Insert:
                                    if (IsLimitEntryOrder(orderRow) && string.IsNullOrEmpty(mOrderID))
                                    {
                                        mOrderID = orderRow.OrderID;
                                        Console.WriteLine("The order has been added. Order ID: {0}, Rate: {1}, Time In Force: {2}",
                                            orderRow.OrderID,
                                            orderRow.Rate,
                                            orderRow.TimeInForce);
                                        mSyncResponseEvent.Set();
                                    }
                                    break;
                                case O2GTableUpdateType.Delete:
                                    Console.WriteLine("The order has been deleted. Order ID: {0}", orderRow.OrderID);
                                    mSyncResponseEvent.Set();
                                    break;
                            }
                        }
                    }
                }
            }
        }

        #endregion

        private bool IsLimitEntryOrder(O2GOrderRow order)
        {
            return order.Type.StartsWith("LE");
        }


    }
}

